pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -1665,73 +1665,95 @@ private void parseQuery() {
+
     for (Map.Entry<String, Object> entry : updateValues.entrySet()) {
       int columnIndex = findColumn(entry.getKey()) - 1;
-
       Object valueObject = entry.getValue();
-      if (valueObject instanceof PGobject) {
-        String value = ((PGobject) valueObject).getValue();
-        rowBuffer.set(columnIndex, (value == null) ? null : connection.encodeString(value));
-      } else {
-        switch (getSQLType(columnIndex + 1)) {
-
-          // boolean needs to be formatted as t or f instead of true or false
-          case Types.BIT:
-          case Types.BOOLEAN:
-            rowBuffer.set(columnIndex, connection
-                .encodeString(((Boolean) valueObject).booleanValue() ? "t" : "f"));
-            break;
-            //
-            // toString() isn't enough for date and time types; we must format it correctly
-            // or we won't be able to re-parse it.
-            //
-          case Types.DATE:
-            rowBuffer.set(columnIndex, connection
-                .encodeString(
-                    connection.getTimestampUtils().toString(
-                        getDefaultCalendar(), (Date) valueObject)));
-            break;
-
-          case Types.TIME:
-            rowBuffer.set(columnIndex, connection
-                .encodeString(
-                    connection.getTimestampUtils().toString(
-                        getDefaultCalendar(), (Time) valueObject)));
-            break;
-
-          case Types.TIMESTAMP:
-            rowBuffer.set(columnIndex, connection.encodeString(
-                connection.getTimestampUtils().toString(
-                    getDefaultCalendar(), (Timestamp) valueObject)));
-            break;
-
-          case Types.NULL:
-            // Should never happen?
-            break;
-
-          case Types.BINARY:
-          case Types.LONGVARBINARY:
-          case Types.VARBINARY:
-            if (isBinary(columnIndex + 1)) {
-              rowBuffer.set(columnIndex, (byte[]) valueObject);
-            } else {
-              try {
-                rowBuffer.set(columnIndex,
-                    PGbytea.toPGString((byte[]) valueObject).getBytes("ISO-8859-1"));
-              } catch (UnsupportedEncodingException e) {
-                throw new PSQLException(
-                    GT.tr("The JVM claims not to support the encoding: {0}", "ISO-8859-1"),
-                    PSQLState.UNEXPECTED_ERROR, e);
-              }
-            }
-            break;
+      setRowBufferColumn(columnIndex, valueObject);
+    }
 
-          default:
-            rowBuffer.set(columnIndex, connection.encodeString(String.valueOf(valueObject)));
-            break;
-        }
+    if (isInsert) {
+      final ResultSet generatedKeys = insertStatement.getGeneratedKeys();
+      try {
+        generatedKeys.next();
+
+        int numKeys = primaryKeys.size();
 
+        for (int i = 0; i < numKeys; i++) {
+          final PrimaryKey key = primaryKeys.get(i);
```

Comment:
The primary keys and their location are not going to change. I don't think this has to be inside this loop.

Reply:
Is it worthwhile to store the index somewhere else? Or did you have something else in mind.

Reply:
ah, right. I need to review what I was trying to get at.

Reply:
Ya, I think we can cache it inside the Result Set it only needs to be retrieved once per result set.

Reply:
I think I can have a look at this tomorrow. But is it really worth it to precompute this? This array is probably going to have a length of 1-3 for most applications. 
For this change to make sense, I guess the two other loops over primaryKeys have to be changed as well.

Reply:
If I'm not mistaken this gets calculated for every time you move to a new  row ? If so it is a big deal

Reply:
We should only hit this when performing an insert of the new row. At this point we are on the way to construct the insert statement.

Reply:
Actually it's called on pretty much every insert, update, move, etc. However it's only called once per result set to I'm fine not caching it.